hdfs/src/main/scala/akka/stream/alpakka/hdfs/impl/HdfsFlowStage.scala

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.hdfs.impl
+
+import akka.NotUsed
+import akka.event.Logging
+import akka.stream.alpakka.hdfs._
+import akka.stream.alpakka.hdfs.impl.HdfsFlowLogic.{FlowState, FlowStep, LogicState}
+import akka.stream.alpakka.hdfs.impl.strategy.DefaultRotationStrategy.TimeRotationStrategy
+import akka.stream.alpakka.hdfs.impl.writer.HdfsWriter
+import akka.stream.stage._
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import cats.data.State
+
+/**
+ * Internal API
+ */
+private[hdfs] final class HdfsFlowStage[W, I, C](
+    ss: SyncStrategy,
+    rs: RotationStrategy,
+    settings: HdfsWritingSettings,
+    hdfsWriter: HdfsWriter[W, I]
+) extends GraphStage[FlowShape[IncomingMessage[I, C], OutgoingMessage[C]]] {
+
+  private val in = Inlet[IncomingMessage[I, C]](Logging.simpleName(this) + ".in")
+  private val out = Outlet[OutgoingMessage[C]](Logging.simpleName(this) + ".out")
+
+  override val shape: FlowShape[IncomingMessage[I, C], OutgoingMessage[C]] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new HdfsFlowLogic(ss, rs, settings, hdfsWriter, in, out, shape)
+}
+
+/**
+ * Internal API
+ */
+private final class HdfsFlowLogic[W, I, C](
+    initialSyncStrategy: SyncStrategy,
+    initialRotationStrategy: RotationStrategy,
+    settings: HdfsWritingSettings,
+    initialHdfsWriter: HdfsWriter[W, I],
+    inlet: Inlet[IncomingMessage[I, C]],
+    outlet: Outlet[OutgoingMessage[C]],
+    shape: FlowShape[IncomingMessage[I, C], OutgoingMessage[C]]
+) extends TimerGraphStageLogic(shape)
+    with InHandler
+    with OutHandler {
+
+  private var state = FlowState(initialHdfsWriter, initialRotationStrategy, initialSyncStrategy)
+
+  private val flushProgram = rotateOutput.flatMap(message => tryPush(Seq(message)))
+
+  setHandlers(inlet, outlet, this)
+
+  def onPush(): Unit =
+    state = onPushProgram(grab(inlet))
+      .runS(state)
+      .value
+
+  def onPull(): Unit =
+    tryPull()
+
+  override def preStart(): Unit = {
+    // Schedule timer to rotate output file
+    initialRotationStrategy match {
```

Comment:
This should be expressed by something in `RotationStrategy` instead of the type so it becomes extendible.

Reply:
Do you have any hint about this? How can i have a strategy in RotationStrategy for preStart. I would consider to pass stateLogic as a parameter and call schedule there but these methods are not visible. How can i trigger schedule in RotationStrategy.

Reply:
You could add an `interval: Option[FiniteDuration]` to it and use `foreach` in pre-start to schedule the poll it.